add --output-format=json output option to v2 list

cosmicexplorer opened the pull request:
### Problem

Resolves #8445.

### Solution

- Add an `--output-format` option to the `list` v2 `@console_rule`, and make `--provides` and `--documented` point to values of that enum option.
- Add `--output-format=json`, which prints out lines of json with the keys:
  - `was_root`: whether the target was a target root, or one of the transitive dependencies of one of the roots
  - `address`: the target address
  - `target_type`: the stringified version of the target's BUILD file name, e.g. `python_library`
  - `intransitive_fingerprint`: the intransitive fingerprint for that `TargetAdaptor`
  - `transitive_fingerprint`: the transitive fingerprint for that `TargetAdaptor`

### Result

The following command line will output a string representing a stable hash of the transitive closure of the target `my/python:binary`:
```bash
$ ./pants list --output-format=json my/python:binary \
  | jq -r 'select(.was_root) | .transitive_fingerprint'
ef54aa0c26c8d91bb74cd575e0cac9378bef8e4a
```

Once #7356 lands, we can use the following command to print the fingerprints of all python_binary targets whenever their source files change:
```bash
$ ./pants --loop --query="type_filter('python_binary')" list --output-format=json :: \
  jq 'select(.was_root) | {.address, .transitive_fingerprint}'
{"address": "my/python:binary", "transitive_fingerprint": "ef54aa0c26c8d91bb74cd575e0cac9378bef8e4a"}
```

cosmicexplorer commented:
Note that I didn't discuss this specific implementation with folks earlier and would love to hear input on this approach.

cosmicexplorer commented:
This is likely to be extremely useful when running pants with `--loop` for scalameta/metals#935, according to the author of that PR.

cosmicexplorer commented:
I absolutely love `--output=json` or maybe `--output-format=json`!

cosmicexplorer commented:
(Ideally that sets the stage for a v2 `./pants export` too!)

cosmicexplorer commented:
Done! Made an enum `--output-format` option, defaulting to `--output-format=address-specs`!

cosmicexplorer commented:
It's correct that this previously calculated just the hashes of the fields. It's not clear to me why the sources field is excluded from calculation in e.g. `Struct._key()`, but this implementation I've just pushed will explicitly attempt to extract sources from targets.

illicitonion commented:
Stepping-back question: What's this actually useful for, if it doesn't include the digests of source files?

(FWIW it would now be trivial to mix in a source-file digest by mixing-in `TargetAdaptor.sources.snapshot`)

illicitonion commented:
> Stepping-back question: What's this actually useful for, if it doesn't include the digests of source files?
> 
> (FWIW it would now be trivial to mix in a source-file digest by mixing-in `TargetAdaptor.sources.snapshot`)

Oops, I was just looking at the comments here not the code - you're already doing this! :)

blorente commented on src/python/pants/engine/legacy/graph.py:
```diff
@@ -511,19 +519,44 @@ def transitive_fingerprinted_targets(
+    for dep in tht.dependencies:
+      dep_intransitive_fingerprint = intransitive_fingerprint_dict.get(dep.root.address, None)
+      if not dep_intransitive_fingerprint:
+        dep_sources = getattr(dep.root.adaptor, 'sources', None)
```
Nit: This block and the one before looks like it could be extracted into a common function

cosmicexplorer replied:
Done now via `@memoized_classproperty!

illicitonion commented on src/python/pants/engine/legacy/structs.py:
```diff
@@ -29,6 +29,20 @@ class TargetAdaptor(StructWithDeps):
+    # `stable_json_sha1()` to fail with a cycle detection. Since some python targets are only mapped
+    # to `TargetAdaptor` (and not `PythonTargetAdaptor`), we check every single target for a
+    # `requirements` kwarg, which is fine for now.
+    key, value = super()._coerce_key_values(key, value)
```
I'm not sure from reading - does this cover all actual key-values of the `Target`, or just the ones explicitly listed in BUILD files?

In particular, if we change a default value in pants, or set a default with a flag or something, and the target doesn't set it, will the fingerprint change? Presumably it _should_, right?

cosmicexplorer replied:
This covers all the key-values that are provided as `field_adaptors`, I believe, essentially because we're getting this info from a `TargetAdaptor`, which will explicitly only use `HydrateableField` `@union` members as the `_kwargs` provided by `Struct` -- see `hydrate_targets`: https://github.com/pantsbuild/pants/blob/1988587507cb2cb21efffac01ac9a8c3ebf291a5/src/python/pants/engine/legacy/graph.py#L497-L509

> In particular, if we change a default value in pants, or set a default with a flag or something, and the target doesn't set it, will the fingerprint change? Presumably it *should*, right?

Short answer: no, this fingerprint will not necessarily change, and yes, I absolutely think it should before we merge this.

Longer answer: Not all the things that contribute to `Target#fingerprint` do _not_ transfer to a `TargetAdaptor` (which subclasses `StructWithDeps`, which subclasses `Struct`), *only* the things which are marked as hydrateable fields. As you imply, this is possibly not what we want at all for this purpose, and my use of `TargetAdaptor` here might not be correct.

@stuhood do you have any insight on how to bridge this? Does it break the v2 build graph traversal model if we're able to get an instance of a real `Target` in order to get a more normal fingerprint (i.e. a fingerprint containing everything the target payload does)? Are the `Payload`/`Target` concepts necessarily v1-only/requiring a v1 build graph, or is it possible to avoid having to reimplement payloads for all targets? I would love to pair on this or make an issue as necessary.

illicitonion replied:
Can we remove the fingerprints from the output before merging this?
It feels like target_type is an obviously useful thing to include; it's less obvious that these fingerprints are things we should be exposing as-is, but we can always add them in the future if we need to / firm them up :)

cosmicexplorer commented on tests/python/pants_test/backend/graph_info/tasks/test_list_targets.py:
```diff
@@ -178,3 +179,104 @@ def test_list_documented(self):
+        "was_root": True,
+        "address": "f:alias",
+        "target_type": "target",
+        "intransitive_fingerprint": "c108686e1fc1b327af1dbc295008762559d0b410",
```
Yes, I agree to all of the above, and would absolutely prefer not to add new testing that blindly tries to match fingerprints. I will add this testing!

ShaneDelmore commented:
+1 I could use this for the work I am currently doing.  The addition of target type to list is particularly useful.

ShaneDelmore commented:
An additional attribute of targets that would be useful, but that I would not block the PR for, is internal/external.

cosmicexplorer commented:
> internal/external

`./pants list` as we've implemented here only lists target roots -- are we looking to list dependencies as well?

stuhood commented:
Honest question: at this point, should we be leaning in to provide richer information via `query`, and keeping `list` simpler?

cosmicexplorer commented:
`--query` is great, but I wasn't thinking that `--query` would do the job of *presenting* information about targets (which this PR does), just *selecting* them. For example, `--output-format=json` can effectively be used with `--query` and `--loop` to e.g. get the fingerprint of a target whenever its source files change.

cosmicexplorer commented:
I've also made the tests a lot simpler by just comparing the changed fingerprint values instead of trying to add a specific fingerprinting mechanism / fingerprint value to the pants public API.

cosmicexplorer commented:
@ShaneDelmore I've just thought back to your question and I think this PR has the information you're looking for: this is one of the keys in the json output with `--output-format=json`:
>   - `was_root`: whether the target was a target root, or one of the transitive dependencies of one of the roots

ShaneDelmore commented:
> > internal/external
> 
> `./pants list` as we've implemented here only lists target roots -- are we looking to list dependencies as well?

I have no need for dependencies in list. A way to list all targets (artifacts?  libraries?  not sure of terminology) used and whether they are internal or external would be handy.  Worst case I could compare pants list to pants depmap and any items depended on that aren't in list must be external right?

ShaneDelmore commented:
> @ShaneDelmore I've just thought back to your question and I think this PR has the information you're looking for: this is one of the keys in the json output with `--output-format=json`:
> 
> > * `was_root`: whether the target was a target root, or one of the transitive dependencies of one of the roots

What is the definition of target root? I thought it was the output of filter-minimize, the targets you needed to invoke to get all code compiled eventually, but based on your usage I'm thinking I was wrong and it is actually all targets defined in BUILD.* files maybe.

cosmicexplorer commented:
> What is the definition of target root? I thought it was the output of filter-minimize, the targets you needed to invoke to get all code compiled eventually, but based on your usage I'm thinking I was wrong and it is actually all targets defined in BUILD.* files maybe.

"target roots" are all targets on the command line. `filter-minimize` (which is a v1 task that subclasses `Filter` in the twitter monorepo) mimics that to obtain the smallest number of target roots. that's useful for us because we can then execute each target in its own pants invocation, which twitter uses in its internal CI -- that reduces the number of separate parallel pants invocations we know we have to invoke via https://github.com/twitter/scoot.

As a note, I've added a `minimize()` method to #7356, so `filter-minimize` should be available for the general public soon too.

cosmicexplorer commented:
This failed on unrelated test timeouts and a mypy check which is now fixed, so it is definitely green. Since @ShaneDelmore and @olafurpg have expressed interest in this feature and I don't know of an obvious alternative, I would love to merge this change. @stuhood I have modified the tests to avoid relying on any specific fingerprinting mechanism, which means that no code should be relying on these fingerprints, or the structure of this json output, being the same across pants versions *(yet)*. Would that guarantee be sufficient to overcome your concern about investing further in `list` vs `--query`?

cosmicexplorer commented:
Going to merge this if there are no further comments in the next day or so since there is a clear user need and some review has been performed.

cosmicexplorer commented:
Splitting this into two PRs to separate:
(1) refactoring `list_targets.py`
(2) adding `TransitiveFingerprintedTargets` (and conform to the new v2 target api!)

cosmicexplorer commented:
Noting that the buck build tool supports a “show target hash” option for a while, which has allowed hooking it up to a system containing bazel. It would have been really great to have had less pushback on this PR initially when there was a clearly present user need.

stuhood commented:
See the discussion in the linked ticket for more information on why this is subtle: https://github.com/bazelbuild/bazel/issues/7962 ... file digest and action digests are useful for very different things: it's really important that users don't think that this is the latter thing (ie, they will need to do their own digesting of all of pant's other config, etc).

I've suggested in slack that I think a `RuleGraph`-aware `query` might allow for exposing more of the subtlety here, because it has the potential to allow for querying the digest of a particular goal, or of the inputs to a particular process (which would include all of the kinds of config you need in an action graph fingerprint).

Given that that might be a ways away though, I think that we could move forward here if we resolve a few things:
1. the naming of the properties: we should make sure it is clear that they only include file contents (ie, the digest will not change if pants' configuration changes): maybe `files_only_fingerprint`?
2. by default, `list` does not require hydrating targets, or walking into their transitive deps: both of those add a non-trivial cost, so the `json` output should probably not contain all of those properties by default. So the json should maybe further allow for field filtering, with digests disabled by default.

benjyw commented:
`content_fingerprint` ?

stuhood commented:
> `content_fingerprint` ?

Probably not specific enough... content of what? But for that matter, `files_only` is not very clear either, heh. Not sure what to call this.

Eric-Arellano commented:
FYI https://github.com/pantsbuild/pants/pull/9912 will impact this, hopefully making things easier. We go back to having only one `list` implementation, this time using the Target API.

Eric-Arellano commented:
Closing as stale, which we're doing for all changes that haven't been touched in 1+ years to simplify project management.

This would still be a really neat feature, though. Do feel free to reopen. Thank you for showing what `--query` could look like for Pants!
